packages/x-data-grid-pro/src/hooks/features/dataSource/cache.ts

```diff
@@ -1,50 +1,132 @@
 import { GridGetRowsParams, GridGetRowsResponse } from '../../../models';
 
-type GridDataSourceCacheDefaultConfig = {
+export type GridDataSourceCacheDefaultConfig = {
   /**
    * Time To Live for each cache entry in milliseconds.
    * After this time the cache entry will become stale and the next query will result in cache miss.
    * @default 300000 (5 minutes)
    */
   ttl?: number;
+  /**
+   * The number of rows to store in each cache entry. If not set, the whole array will be stored in a single cache entry.
+   * Setting this value to smallest page size will result in better cache hit rate.
+   * Has no effect if cursor pagination is used.
+   * @default undefined
+   */
+  chunkSize?: number;
 };
 
 function getKey(params: GridGetRowsParams) {
   return JSON.stringify([
-    params.paginationModel,
     params.filterModel,
     params.sortModel,
     params.groupKeys,
+    params.start,
+    params.end,
   ]);
 }
 
 export class GridDataSourceCacheDefault {
-  private cache: Record<string, { value: GridGetRowsResponse; expiry: number }>;
+  private cache: Record<
+    string,
+    {
+      value: GridGetRowsResponse;
+      expiry: number;
+      chunk: { startIndex: string | number; endIndex: number };
+    }
+  >;
 
   private ttl: number;
 
-  constructor({ ttl = 300000 }: GridDataSourceCacheDefaultConfig) {
+  private chunkSize: number;
+
+  private getChunkRanges = (params: GridGetRowsParams) => {
```

Comment:
I'm wondering what should be the behavior for the users using custom caches, they'll have to replicate this chunking logic in their cache abstraction, right?
Would it be better to extract the chunking logic somewhere between the Datasource layer and the caching layer?
IMO the default cache should not contain business logic, just the basic getter, setter, and invalidator. That way it'd be easier to support optimizations like chunking for the custom caches.

I tried to visualize it in this diagram.

![image](https://github.com/user-attachments/assets/c526437b-30a7-4791-8095-198b5c5ce00b)

What do you think?

Reply:
users don't have to replicate it
they can implement cache anyway they like. it might get better/worse hit ratio, but it will not break things.

I can extract the logic to make the code cleaner, but I would not (at least not at this point) expose any of these publicly.

There might be some issues with it once it is out and it will be easier to fix things if we don't have to worry about breaking changes. Once stable, we can open these pieces so that users can build on top of it

initially it will be a choice between:
- no cache
- our cache
- custom developed cache

Reply:
>  expose any of these publicly

I don't think we need to expose it publically. It (let's say `ChunksManager`) would just be an internal module like the data source, whose responsibilities will be:
1. To divide the currently fetched page range into chunks for storing in cache
	- Example: range: `{start: 1, end: 20}` chunks: `{start: 1, end: 10}, {start: 11, end: 20}`
2. To compute chunks that should be searched in the cache to check if the required data is available in cache.
	- Example: requested range: `{start: 5, end: 15}`  chunks to search: `{start: 1, end: 10}, {start: 11, end: 20}`

The aim is to have same hit ratio for internal cache and users' caches, in other case, if IIUC, the pages which require chunking would never be cached on custom cache unless the user support a similar chunking logic in the cache. Am I missing something?

Reply:
We can have this `ChunksManager` enabled by default even if lazy loading is not enabled ([because of](https://github.com/mui/mui-x/pull/13878#discussion_r1797314764))
Should we allow users to disable this if they don't want it?

Reply:
> Should we allow users to disable this if they don't want it?

Do you have any use cases for this in mind? If not, I don't think we should do it at this point.

Bilal's point makes sense to me though – chunk manager should be a separate entity, not part of the cache. Our default cache can still use it though.
In the future, we can expose chunk manager to allow using it in a custom cache.

Reply:
> Our default cache can still use it though.
In the future, we can expose chunk manager to allow using it in a custom cache.

I intended to make the Chunks Manager a separate entity but also to not make it an optional thing that we can use in a cache and expose later to be used by users. Rather it'd be an integral part of the Data Source logic. Whenever some data is fetched from the server (point 1 below) or retrieved from the cache (point 2 below), the Data Grid will use Chunks Manager to check:
1. Which chunks a specific page should be divided into for saving in the cache?
2. Which chunks to search to get a specific page from the cache?

That way it will be independent of the cache layer itself, it will sit between the data source and the cache as an Adapter (as I _tried_ to explain in the above diagram, but it might be not very clear)

Does that make sense?

Reply:
Code is split. Data source uses the new chunk manager by default. I will extend the docs explaining how the caching works under the hood

Reply:
Documented here
https://deploy-preview-13878--material-ui-x.netlify.app/x/react-data-grid/server-side-data/#improving-the-cache-hit-ratio